[FLINK-11159] Allow configuration whether to fall back to savepoints for restore

yanghua opened the pull request:
## What is the purpose of the change

*This pull request provides a checkpoint config option `isPreferCheckpointForRecovery` to allow configuration whether to fall back to savepoints for restore*

## Brief change log

  - *Provide a checkpoint config option `isPreferCheckpointForRecovery`* 
  - *Allow configuration whether to fall back to savepoints for restore*
  - *Add two test cases*

## Verifying this change


This change added tests and can be verified as follows:

  - *testRestoreLatestCheckpointWhenPreferCheckpoint*
  - *testRestoreLatestCheckpointWhenPreferSavepoint*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

yanghua commented:
@gyfora If you have time, can you review this PR for me? thanks.

gyfora commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -183,6 +184,8 @@
 	/** Registry that tracks state which is shared across (incremental) checkpoints. */
 	private SharedStateRegistry sharedStateRegistry;
 
+	private boolean isPerferCheckpointForRecovery;
```
There is a typo in basically every use of the word "preferCheckpointForRecovery" it should be prefer not "perfer". I think the easiest would be to search the changes and correct it in every place.

yanghua replied:
Sorry for the typo, will fix it soon.

gyfora commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -1093,6 +1098,23 @@ public boolean restoreLatestCheckpointedState(
 			}
 
+			CompletedCheckpoint candidate = null;
+			if (isPerferCheckpointForRecovery && completedCheckpointStore.getAllCheckpoints().size() > 1) {
```
It feels like this part basically reimplements the getLatestCheckpoint() logic of the completed checkpoint store. Maybe instead of having this hear the getLatestCheckpoint( ) could take the preferCheckpoints boolean parameter for better encapsulation

yanghua replied:
Agree.

gyfora commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java:
```diff
@@ -72,6 +72,9 @@
 	private boolean failOnCheckpointingErrors = true;
 
+	/** Determines if a job will fallback to checkpoint when there is a savepoint later than checkpoint. **/
+	private boolean perferCheckpointForRecovery = false;
```
I am personally not a big fan of configuring checkpoint parameters in the CheckpointConfig class, I actually wish that all these options would be exposed as general flink configuration parameters (settable in flink-conf.yaml).

What do you think @StefanRRichter ?

yanghua replied:
I think this config option belongs to checkpoint field. So provide it in `CheckpointConfig` sounds more reasonable. What's more, it is a job level config option, right? However, `flink-conf.yaml` is a cluster-level configuration file(if users deploy Flink with standalone mode, it will affect all the jobs).

gyfora commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java:
```diff
@@ -285,6 +288,16 @@ public boolean isExternalizedCheckpointsEnabled() {
 
+	@PublicEvolving
+	public boolean isPreferCheckpointForRecovery() {
+		return perferCheckpointForRecovery;
```
There are still a few typos if you search for "perfer"

yanghua replied:
Sorry, will fix them.

gyfora commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StandaloneCompletedCheckpointStore.java:
```diff
@@ -77,8 +79,32 @@ public void addCheckpoint(CompletedCheckpoint checkpoint) throws Exception {
+			return null;
+		}
+
+		CompletedCheckpoint candidate = checkpoints.getLast();
```
Do you think we could somehow avoid having pretty much exactly the same logic duplicated twice (ZK/Standalone)? Maybe the whole logic could be implemented in the CompletedCheckpointStore interface based on getAllCheckpoints()

yanghua replied:
Agree, actually, I found the similar code snippet when I was implementing. The difference between the original implementation of the `getLatestCheckpoint` is the usage of `peekLast` and `getLast`. I was afraid to destroy the original semantics. If it does not matter, I'd like to do this refactor.

gyfora replied:
we should preserve the current behaviour, but the difference between getLast and peekLast is minor, just check for empty list I guess

yanghua replied:
I will give a default implementation in the `CompletedCheckpointStore `:

```java
default CompletedCheckpoint getLatestCheckpoint(boolean isPreferCheckpointForRecovery) throws Exception {
		if (getAllCheckpoints().isEmpty()) {
			return null;
		}

		CompletedCheckpoint candidate = getAllCheckpoints().get(getAllCheckpoints().size() - 1);
		if (isPreferCheckpointForRecovery && getAllCheckpoints().size() > 1) {
			List<CompletedCheckpoint> allCheckpoints;
			try {
				allCheckpoints = getAllCheckpoints();
				ListIterator<CompletedCheckpoint> listIterator = allCheckpoints.listIterator(allCheckpoints.size() - 1);
				while (listIterator.hasPrevious()) {
					CompletedCheckpoint prev = listIterator.previous();
					if (!prev.getProperties().isSavepoint()) {
						candidate = prev;
						LOG.info("Found a completed checkpoint before the latest savepoint, will use it to recover!");
						break;
					}
				}
			} catch (Exception e) {
				LOG.error("Method getAllCheckpoints caused exception : ", e);
				throw new FlinkRuntimeException(e);
			}
		}

		return candidate;
	}
```

gyfora replied:
Sounds good, as long as we can be sure getAllCheckpoints() returns checkpoint in order.

R-CodeBoy replied:
Does it mean that the latest checkpoint is be surely  in the last position in list? thanks

yanghua replied:
@R-CodeBoy Yes, the `getAllCheckpoints ` returns an ArrayList which wrapped an ArrayDeque.

R-CodeBoy replied:
@yanghua Do you compare the cost time of recovery between checkpoint and savepoint,In my options ,sometimes it depends on the checkpoint interval when considering the recovery speed or reprocessing the data.

yanghua replied:
Actually, it is a config option which depends on users themselves production scene. Some scene, the cost of rolling offset back is higher than recovery time, however, some scene is converse. To guess the scene is difficult to a general compute engine.

yanghua commented:
@gyfora What do you think about the new version of this PR?

yanghua commented:
It's very strange, the Travis reported compile error :

```
16:40:05.120 [ERROR] /home/travis/build/apache/flink/flink-runtime/src/test/java/org/apache/flink/runtime/jobgraph/JobGraphTest.java:[343,95] constructor CheckpointCoordinatorConfiguration in class org.apache.flink.runtime.jobgraph.tasks.CheckpointCoordinatorConfiguration cannot be applied to given types;
16:40:05.120 [ERROR] required: long,long,long,int,org.apache.flink.runtime.checkpoint.CheckpointRetentionPolicy,boolean,boolean
16:40:05.120 [ERROR] found: long,long,long,int,org.apache.flink.runtime.checkpoint.CheckpointRetentionPolicy,boolean
```

But the class `JobGraphTest` does not use `CheckpointCoordinatorConfiguration`.

yanghua commented:
cc @gyfora

gyfora commented:
What do you mean it doesnt use it? you can see it here: https://github.com/apache/flink/blob/master/flink-runtime/src/test/java/org/apache/flink/runtime/jobgraph/JobGraphTest.java#L343

yanghua commented:
Oh @gyfora Sorry, yes, I have found it after I gave that comment. The reason is that I did not rebase the latest master code. Now, it works fine. I ping you to do the final check.

gyfora commented:
The changes look good, last thing that would be great to have is some documentation for the new feature :)

yanghua commented:
@gyfora OK. Will supply the docs.

yanghua commented:
@gyfora I have added java doc and flink doc. Please review.

gyfora commented on docs/dev/stream/state/checkpointing.md:
```diff
@@ -76,6 +76,8 @@ Other parameters for checkpointing include:
 
   - *fail/continue task on checkpoint errors*: This determines if a task will be failed if an error occurs in the execution of the task's checkpoint procedure. This is the default behaviour. Alternatively, when this is disabled, the task will simply decline the checkpoint to the checkpoint coordinator and continue running.
 
+  - *is prefer checkpoint for recovery*: This determines if a job will fallback to checkpoint when there is a savepoint later than checkpoint.
```
I would change the wording slightly here to be more grammatically correct:

*prefer checkpoint for recovery*: This determines if a job will fallback to latest checkpoint even when there are more recent savepoints available to potentially reduce recovery time

gyfora commented on docs/dev/stream/state/checkpointing.zh.md:
```diff
@@ -76,6 +76,8 @@ Other parameters for checkpointing include:
 
   - *fail/continue task on checkpoint errors*: This determines if a task will be failed if an error occurs in the execution of the task's checkpoint procedure. This is the default behaviour. Alternatively, when this is disabled, the task will simply decline the checkpoint to the checkpoint coordinator and continue running.
 
+  - *is prefer checkpoint for recovery*: This determines if a job will fallback to checkpoint when there is a savepoint later than checkpoint.
```
I dont know what's the difference between the checkpointing.zh.md and checkpointing.md files, do we need to change both?

yanghua replied:
The files with pattern `xxx.zh.md` is copied for translating purpose. They will be translated into Chinese. So for the correctness, I think change both would be a good solution.

gyfora commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java:
```diff
@@ -285,6 +288,24 @@ public boolean isExternalizedCheckpointsEnabled() {
 	}
 
+	/**
+	 * Returns whether a job recovery should fallback to checkpoint when there is a savepoint later than checkpoint.
```
I would replace in all the docs : " when there is a savepoint later than checkpoint" with "when there is a more recent savepoint"

gyfora commented:
I had a few smaller doc comments, otherwise the changes look pretty good. I dont know how the different doc files are handled so once thats cleared up and we squash the commits I think this is good to go :)

yanghua commented:
@gyfora Thanks for correcting some representation problems. I will update the documentation. And I want to make sure that if I need to squash these commits or you would do it?

gyfora commented:
Once the content is correct you either squash it or I will do it myself when I merge, its up to you :) I think we can merge this today

yanghua commented:
@gyfora Thanks. I have refactored the java doc and flink doc based on your suggestion. In addition, I have also squashed commits and rebased the branch.

StephanEwen commented:
This PR unfortunately changes a few sensitive core components  (CheckpointCoordinator and CompletedCheckpointStore) and introduces a new configuration while pretty much violating all best practices and code style of Flink :-(

Just some examples from a quick look:
  - constant fields should be final in class
  - exception rethrown as RuntimeExceptions (when signature even allows exceptions)
  - mockito for testing
  - operational config in ExecutionConfig rather than in flink-conf.yaml
  - bad logging `LOG.error("Method getAllCheckpoints caused exception : ", e);`
  - default method used to misuse interface as abstract class
  - interface level logger instance (side effect of the above)
  - and various other things in the details...

I am inclined to push for reverting this commit and reopen the issue for another implementation attempt. The amount of code issues are too large for a patchup here, in my opinion.

Also, as a personal take: I think we need to put an emphasis on setting good examples for the community to improve the code quality in Flink. If we don't set good examples and gradually improve the code quality, Flink will be harder and harder to maintain in the future.

zentol commented:
The CompletedCheckpointStore is indeed quite sketchy; there are better ways to share logic ...

The mocking is _less_ problematic in my view given that the test files we're modifying are already filled to the brim with mocking (and IIRC the checkpoint coordinator is notoriously difficult to interact with in tests without mocking).
Naturally I'd prefer if we'd rework these tests entirely, but we can't just put this burden on the next guy to touch these tests. That said there are quite a few places though where we could mocking though this in simple ways. (e.g., there's no reason to mock a StreamStatehandle)

It also seems like the documentation is rather light on potential downsides to using this feature, which were discussed in the JIRA.

There's quite a bit of cleanup to be done here, as such I generally wouldn't oppose reverting it. However, there's a time constraint in play with the feature freeze coming up. Would we be fine for delaying this feature? If not, should we find a committer first to either review a cleaned up version, or even do the cleanup itself, before reverting it?

gyfora commented:
@StephanEwen @zentol I agree with most comments, and as far as the logging, exception handling goes it should have been reviewed better.

As for the comment related to operational config in ExecutionConfig. I think the main problem is the existence of CheckpointConfig. Most of the checkpointing related settings can only be configured there at the moment so it doesnt make sense to put something new only in flink-conf. That would be extremely confusing. I think we need to rework the whole CheckpointConfig logic to allow every single thing to be override from flink-conf.yaml.

I was not aware that we are trying to avoid interface default methods in the Flink codebase, I am sorry. I think they can make code much cleaner by removing the need to introduce abstract classes everywhere.

Overall I agree that we need to clean this up one way or the other, I am not opposed to reverting this before the 1.9.0 release.
